ooniprobe/Model/Database/Measurement.mm

```diff
@@ -96,4 +96,35 @@ -(void)deleteObject{
     [self remove];
 }
 
+-(void)getExplorerUrl:(void (^)(NSString*))successcb error:(void (^)(NSError*))errorcb {
+    NSLog(@"%@ getExplorerUrl",self.Id);
+    NSMutableString *path = [NSMutableString stringWithFormat:@"https://api.ooni.io/api/v1/measurements?report_id=%@",
+                      self.report_id];
+    if ([self.test_name isEqualToString:@"web_connectivity"])
+        [path appendString:[NSString stringWithFormat:@"&input=%@",
+                            self.url_id.url]];
+    NSURL *url = [NSURL URLWithString:path];
+    NSURLSessionDataTask *downloadTask =
+    [[NSURLSession sharedSession]
+     dataTaskWithURL:url
+     completionHandler:^(NSData *data, NSURLResponse *response, NSError *error) {
+         if (error != nil)
+             errorcb(error);
+         else {
+             NSDictionary *dic = [NSJSONSerialization JSONObjectWithData:data options:0 error:&error];
+             if (error != nil)
+                 errorcb(error);
+             NSArray *resultsArray = [dic objectForKey:@"results"];
+             //TODO symbolize somehow empty array
+             if ([resultsArray count] > 0)
+                 errorcb([NSError errorWithDomain:@"io.ooni.api"
+                                           code:100
+                                       userInfo:@{NSLocalizedDescriptionKey:@"Error.JsonEmpty"
+                                                  }]);
+             successcb([[resultsArray objectAtIndex:0] objectForKey:@"measurement_url"]);
```

Comment:
This code is still not working as intended. Please, apply this pattern:

```ObjectiveC
{
  completionHandler:^(NSData *data, NSURLResponse *response, NSError *error) {
    if (error != nil) {
      errorcb(error);
      return;
    }
    NSDictionary *dic = [NSJSONSerialization JSONObjectWithData:data options:0 error:&error];
    if (error != nil) {
      errorcb(error);
      return;
    }
    NSArray *resultsArray = [dic objectForKey:@"results"];
    //TODO symbolize somehow empty array
    if ([resultsArray count] <= 0) {
        errorcb([NSError errorWithDomain:@"io.ooni.api"
                                                          code:100
                                                     userInfo:@{NSLocalizedDescriptionKey:@"Error.JsonEmpty"
        }]);
        return;
    }
    successcb([[resultsArray objectAtIndex:0] objectForKey:@"measurement_url"]);
}
```

Also, the check on the array count seems to be inverted.

Reply:
This pattern is better because the code does not grow out of control because of indenting and you can have several clauses. Also, this is how many programmers expect to see the code. This is also for historical reasons as the compilers often assumes that jumps are not taken. Therefore, historically, writing code like this was more performant (and may even be today, but that's really not the point here). Also, and this is actually much more useful to keep in mind, if you organise code like this, you can understand what happens in the "all good path" by just avoid reading the `if`s and their body.